### PR TITLE
Supports text separation

### PR DESCRIPTION
**Pull request by NoodleSamaChan**
# Pull Request

## Related issue
#526

## What does this PR do?
Added two new settings: separatorTokens and nonSeparatorTokens with get, update, and reset methods associated.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: .code-samples.meilisearch.yaml [hunk 1]
```diff
@@ -839,6 +875,15 @@ search_parameter_guide_show_ranking_score_1: |-
...
     .unwrap();
+search_parameter_guide_show_ranking_score_details_1: |-
+  let results: SearchResults<Movie> = client
+    .index("movies")
+    .search()
+    .with_query("dragon")
+    .with_show_ranking_score_details(true)
+    .execute()
+    .await
+    .unwrap();
```
Comment (irevoire) [inline .code-samples.meilisearch.yaml:886]: This was related to your previous PR.
It’s not supposed to be here

```suggestion
```

File: src/settings.rs [hunk 2]
```diff
@@ -103,6 +103,12 @@ pub struct Settings {
     /// Proximity precision settings.
     #[serde(skip_serializing_if = "Option::is_none")]
     pub proximity_precision: Option<String>,
+    /// Fields in which to search for matching query words sorted by order of importance.
```
Comment (irevoire) [inline src/settings.rs:106]:
```suggestion
    /// Configure strings as custom separator tokens indicating where a word ends and begins.
```

File: src/settings.rs [hunk 3]
```diff
@@ -103,6 +103,12 @@ pub struct Settings {
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub separator_tokens: Option<Vec<String>>,
+    /// Fields in which to search for matching query words sorted by order of importance.
```
Comment (irevoire) [inline src/settings.rs:109]: Please take a look at the documentation for these comments; https://www.meilisearch.com/docs/reference/api/settings#non-separator-tokens

```suggestion
    /// Remove tokens from Meilisearch's default [list of word separators](https://www.meilisearch.com/docs/learn/engine/datatypes#string).
```

File: src/settings.rs [hunk 4]
```diff
@@ -2066,4 +2314,72 @@ mod tests {
...
+            .set_separator_tokens(&vec!["@".to_string(), "#".to_string()])
+            .await
+            .unwrap();
+        client.wait_for_task(task_info, None, None).await.unwrap();
+
+        let mut res = index.get_separator_tokens().await.unwrap();
+        res.reverse();
+
+        assert_eq!(expected, res);
+    }
```
Comment (irevoire) [inline src/settings.rs:2340]:
```suggestion
    #[meilisearch_test]
    async fn test_set_separator_tokens(client: Client, index: Index) {
        let expected: Vec<String> = vec!["@".to_string(), "#".to_string()];

        let task_info = index
            .set_separator_tokens(&expected)
            .await
            .unwrap();
        client.wait_for_task(task_info, None, None).await.unwrap();

        let mut res = index.get_separator_tokens().await.unwrap();

        assert_eq!(expected, res);
    }
```

I believe we should be able to do that

File: src/settings.rs [hunk 5]
```diff
@@ -2066,4 +2314,72 @@ mod tests {
...
+            .set_non_separator_tokens(&vec!["@".to_string(), "#".to_string()])
+            .await
+            .unwrap();
+        client.wait_for_task(task_info, None, None).await.unwrap();
+
+        let mut res = index.get_non_separator_tokens().await.unwrap();
+        res.reverse();
+
+        assert_eq!(expected, res);
+    }
```
Comment (irevoire) [inline src/settings.rs:2374]:
```suggestion
    #[meilisearch_test]
    async fn test_set_non_separator_tokens(client: Client, index: Index) {
        let expected: Vec<String> = vec![ "@".to_string(), "#".to_string()];

        let task_info = index
            .set_non_separator_tokens(&expected)
            .await
            .unwrap();
        client.wait_for_task(task_info, None, None).await.unwrap();

        let mut res = index.get_non_separator_tokens().await.unwrap();

        assert_eq!(expected, res);
    }
```

Same

**Review by irevoire — CHANGES_REQUESTED** [review]
Hey, codewise it’s perfect, but there is a few comments missing here and there.

Also it looks like you have a conflict with main

File: .code-samples.meilisearch.yaml [hunk 6]
```diff
@@ -840,15 +876,6 @@ search_parameter_guide_show_ranking_score_1: |-
...
     .unwrap();
-search_parameter_guide_show_ranking_score_details_1: |-
-  let results: SearchResults<Movie> = client
-    .index("movies")
-    .search()
-    .with_query("dragon")
-    .with_show_ranking_score_details(true)
-    .execute()
-    .await
-    .unwrap();
```
Comment (irevoire) [inline .code-samples.meilisearch.yaml:851]: Ok my bad, I was confused with the conflicts, you were not supposed to delete this chunk of code

File: src/settings.rs [hunk 7]
```diff
@@ -799,6 +837,67 @@ impl<Http: HttpClient> Index<Http> {
             )
             .await
     }
+    /// Get [separator token](https://www.meilisearch.com/docs/reference/api/settings#separator-tokens) of the [Index].
```
Comment (irevoire) [inline src/settings.rs:840]:
```suggestion

        /// Get [separator token](https://www.meilisearch.com/docs/reference/api/settings#separator-tokens) of the [Index].
```

File: src/settings.rs [hunk 8]
```diff
@@ -1892,7 +2073,7 @@ impl<Http: HttpClient> Index<Http> {
             .await
     }
 
-    /// Reset [search cutoff](https://www.meilisearch.com/docs/reference/api/settings#search-cutoff) settings of the [Index].
+    /// Reset [proximity precision](https://www.meilisearch.com/docs/reference/api/settings#
```
Comment (irevoire) [inline src/settings.rs:2076]:
```suggestion
    /// Reset [search cutoff](https://www.meilisearch.com/docs/reference/api/settings#search-cutoff) settings of the [Index].
```

**Review by irevoire — CHANGES_REQUESTED** [review]
Hey, we're almost there, just a few comments remaining

File: .code-samples.meilisearch.yaml (hunk 6 above)
Comment (NoodleSamaChan) [reply .code-samples.meilisearch.yaml]: sorry about that, I believe I recharged everything appropriately

**Review by irevoire — APPROVED** [review]
Thank you,

bors merge

**Pull request merged**